Frameworks/UIKit/UIImage.mm

```diff
@@ -1033,59 +1035,41 @@ - (void)drawInRect:(CGRect)pos {
 */
 - (void)drawInRect:(CGRect)pos blendMode:(CGBlendMode)mode alpha:(float)alpha {
     //  [BUG: Need to honor mode and alpha]
-    CGContextRef cur = UIGraphicsGetCurrentContext();
+    CGContextRef ctx = UIGraphicsGetCurrentContext();
+    CGImageRef img = getImage(self);
 
-    if (!getImage(self)) {
+    if (img == NULL) {
         TraceWarning(TAG, L"m_pImage = NULL!");
         return;
     }
 
-    CGContextSaveGState(cur);
-    CGContextSetBlendMode(cur, mode);
+    CGContextSaveGState(ctx);
+    CGContextSetBlendMode(ctx, mode);
 
-    if (alpha != 1.0) {
+    if (alpha != 1.0f) {
         TraceVerbose(TAG, L"Should draw with alpha");
     }
 
-    if (_scale == 0) {
+    if (_scale == 0.0f) {
         TraceWarning(TAG, L"Scale should be non-zero!");
         return;
     }
 
-    if (_imageInsets.left == 0 && _imageInsets.top == 0 && _imageInsets.right == 0 && _imageInsets.bottom == 0) {
-        CGRect srcRect;
-        CGRect destRect = pos;
+    // Destination rect taken from pos with origin transformed from TL to BL
+    CGRect dstRect;
+    dstRect.origin = { pos.origin.x, (static_cast<float>(CGBitmapContextGetHeight(ctx)) - pos.origin.y - pos.size.height) };
+    dstRect.size = pos.size;
 
-        srcRect.origin.x = 0;
-        srcRect.origin.y = 0;
-        srcRect.size.width = float(getImage(self)->Backing()->Width());
-        srcRect.size.height = float(getImage(self)->Backing()->Height());
+    CGImageBacking* imgBacking = img->Backing();
 
-        //  Invert srcRect
-        srcRect.origin.y = getImage(self)->Backing()->Height() - srcRect.origin.y;
-        srcRect.size.height = -srcRect.size.height;
-        CGContextDrawImageRect(cur, getImage(self), srcRect, destRect);
-    } else {
-        static const CGPatternCallbacks patCB = { 0, drawNinePatchCallBack, NULL };
-        insetInfo ii = { self, makeRect(0, 0, pos.size.width * _scale, pos.size.height * _scale) };
-
-        CGPatternRef pat = CGPatternCreate((void*)&ii,
-                                           makeRect(0, 0, pos.size.width * _scale, pos.size.height * _scale),
-                                           CGAffineTransformMakeScale(_scale, _scale),
-                                           10,
-                                           10,
-                                           kCGPatternTilingConstantSpacing,
-                                           true,
-                                           &patCB);
-
-        CGFloat alpha = 1.0f;
-        CGContextSetFillPattern(cur, pat, &alpha);
-        CGContextFillRect(cur, pos);
-
-        CGPatternRelease(pat);
-    }
+    CGRect srcRect;
+    srcRect.origin = { 0 };
+    srcRect.size = { static_cast<float>(imgBacking->Width()), static_cast<float>(imgBacking->Height()) };
 
-    CGContextRestoreGState(cur);
+    // Draw image and divide into patches if necessary
+    drawPatches(ctx, self, &srcRect, &dstRect);
```

Comment:
Your comment about things being incorrectly scaled isn't exactly correct. An image named <something>@2x.<ext> should have a scale of 2.0, and when rendered should be half the size of logical pixels. Try it on iOS, something like this (photo1.png exists in WOCCatalog):

UIImage* image = [UIImage imageNamed:@"photo1.png"];
image = [UIImage imageWithCGImage:image.CGImage scale:4.0f orientation:UIImageOrientationUp];
image = [image1 resizableImageWIthCapInsets:UIEdgeInsetsMake(50, 50, 50, 50)];

Try drawing this image (from a UIImageView or drawRect) and you'll see that the image caps shrink (the size is the same, but the image is rendered smaller) according to the scale. This is the same behaviour as when using a CALayer with contents = image.CGImage, contentsGravity = kCAGravityCenter and contentsScale = 4.0.